src/libAtomVM/otp_net.c

```diff
@@ -198,59 +230,59 @@ static term nif_net_getaddrinfo(Context *ctx, int argc, term argv[])
+        term addrinfo_map;
+        if (term_is_invalid_term(addrinfo_keys)) {
+            addrinfo_map = term_alloc_map(3, &ctx->heap);
         } else {
-            TRACE("No protocol defined\n");
+            addrinfo_map = term_alloc_map_maybe_shared(3, addrinfo_keys, &ctx->heap);
         }
-
-        if (p->ai_socktype) {
-            term type_atom = globalcontext_make_atom(global, ATOM_STR("\x4", "type"));
-            term type = interop_atom_term_select_atom(type_table, p->ai_socktype, global);
-            term_set_map_assoc(elt, i++, type_atom, term_is_invalid_term(type) ? UNDEFINED_ATOM : type);
-        } else {
-            TRACE("No socket type defined\n");
+        // The inner addr contains a family, port, and addr
+        term addr_atom = globalcontext_make_atom(global, ATOM_STR("\x4", "addr"));
+        term_set_map_assoc(addrinfo_map, 0, family_atom, inet_atom);
+        term_set_map_assoc(addrinfo_map, 1, PORT_ATOM, term_from_int(port));
+        term address = inet_make_addr4(ntohl(((struct sockaddr_in *) p->ai_addr)->sin_addr.s_addr), &ctx->heap);
+        term_set_map_assoc(addrinfo_map, 2, addr_atom, address);
+
+        if (term_is_invalid_term(addrinfo_keys)) {
+            addrinfo_keys = term_get_map_keys(addrinfo_map);
         }
 
-        if (p->ai_addrlen == sizeof(struct sockaddr_in)) {
-            // The inner addr contains a family, port, and addr
-            term addr_atom = globalcontext_make_atom(global, ATOM_STR("\x4", "addr"));
-            term inner_addr = term_alloc_map(3, &ctx->heap);
-            term_set_map_assoc(inner_addr, 0, family_atom, family);
-            term_set_map_assoc(inner_addr, 1, PORT_ATOM, term_from_int(port));
-            term address = make_address(ctx, (struct sockaddr_in *) p->ai_addr);
-            term_set_map_assoc(inner_addr, 2, addr_atom, address);
-
-            // embed the inner_addr, but reference it from both address and addr
-            // for compatibility with OTP
-            term address_atom = globalcontext_make_atom(global, ATOM_STR("\x7", "address"));
-            term_set_map_assoc(elt, i++, address_atom, inner_addr);
-            term_set_map_assoc(elt, i++, addr_atom, inner_addr);
+        if (p->ai_protocol != 0 && p->ai_socktype != 0) {
+            term result_map = make_getaddrinfo_result(&result_keys, p->ai_protocol, p->ai_socktype, addrinfo_map, ctx->global, &ctx->heap);
+            infos = term_list_prepend(result_map, infos, &ctx->heap);
         } else {
-            TRACE("No address defined\n");
+            term tcp_map = make_getaddrinfo_result(&result_keys, IPPROTO_TCP, SOCK_STREAM, addrinfo_map, ctx->global, &ctx->heap);
+            infos = term_list_prepend(tcp_map, infos, &ctx->heap);
+            term udp_map = make_getaddrinfo_result(&result_keys, IPPROTO_UDP, SOCK_DGRAM, addrinfo_map, ctx->global, &ctx->heap);
+            infos = term_list_prepend(udp_map, infos, &ctx->heap);
```

Comment:
Is this okay to do?  Are we just making stuff up?  (I don't really know the answer -- just wondering)

Reply:
I'm not sure if esp-idf is buggy. For a single lookup (e.g. github.com), it returns a single result with proto == 0 and sock_type == 0. macOS returns several entries, one for TCP and one for UDP. So does Erlang/OTP. So this code is about working around this esp-idf behavior.

Reply:
👍🏽